provider/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptClassloadingCache.kt

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.kotlin.dsl.provider
+
+import org.gradle.api.internal.initialization.ClassLoaderScope
+
+import org.gradle.cache.internal.CrossBuildInMemoryCache
+import org.gradle.cache.internal.CrossBuildInMemoryCacheFactory
+
+import org.gradle.internal.Cast
+import org.gradle.internal.classpath.DefaultClassPath
+import org.gradle.internal.hash.HashCode
+
+import org.gradle.kotlin.dsl.support.loggerFor
+
+import java.io.File
+import java.lang.ref.WeakReference
+import javax.inject.Inject
+
+
+internal
+data class LoadedScriptClass<out T>(
+    val compiledScript: CompiledScript<T>,
+    val scriptClass: Class<*>)
+
+
+private
+val logger = loggerFor<KotlinScriptPluginFactory>()
+
+
+internal
+class KotlinScriptClassloadingCache @Inject constructor(cacheFactory: CrossBuildInMemoryCacheFactory) {
+
+    private
+    val cache: CrossBuildInMemoryCache<ScriptCacheKey, LoadedScriptClass<*>> = cacheFactory.newCache()
+
+    fun <T> loadScriptClass(
+        scriptBlock: ScriptBlock<T>,
+        parentClassLoader: ClassLoader,
+        scopeFactory: () -> ClassLoaderScope,
+        compilation: (ScriptBlock<T>) -> CompiledScript<T>
+    ): LoadedScriptClass<T> {
+
+        val key = ScriptCacheKey(scriptBlock.scriptTemplate.qualifiedName!!, scriptBlock.sourceHash, parentClassLoader)
+        val cached = cache.get(key)
+        if (cached != null) {
+            return Cast.uncheckedCast<LoadedScriptClass<T>>(cached)
+        }
+
+        val compiledScript = compilation(scriptBlock)
+
+        logger.debug("Loading {} from {}", scriptBlock.scriptTemplate.simpleName, scriptBlock.displayName)
+        val scriptClass = classFrom(compiledScript, scopeFactory())
+        return LoadedScriptClass(compiledScript, scriptClass).also {
+            cache.put(key, it)
+        }
+    }
+
+    private
+    fun classFrom(compiledScript: CompiledScript<*>, scope: ClassLoaderScope): Class<*> =
+        classLoaderFor(compiledScript.location, scope)
+            .loadClass(compiledScript.className)
+
+    private
+    fun classLoaderFor(location: File, scope: ClassLoaderScope) =
+        scope
+            .local(DefaultClassPath(location))
+            .lock()
```

Comment:
What's with this call to `lock`? Does something need to be unlocked after calling it?

Reply:
Nope. See [ClassLoaderScope.lock()](https://github.com/gradle/gradle/blob/f001402c8b7f404c923d972f1bc524b2de7b3fcd/subprojects/core/src/main/java/org/gradle/api/internal/initialization/ClassLoaderScope.java#L94-L99)

Reply:
I know that this premature locking has caused issues before with the `org.gradle.configureondemand=true` set.
Has this been tested with that?

Related: #311

Reply:
This isn't new code, this PR simply moves it.
But yeah, this is related to #311 and #607